packages/block-library/src/heading/index.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * Server-side rendering of the `core/heading` block.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Adds a wp-block-heading class to the heading blocks without storing
+ * it in the serialized $content.
+ *
+ * @param array $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ *
+ * @return string The content of the block being rendered.
+ */
+function block_core_heading_render( $attributes, $content ) {
+	$matches = array();
+	$pattern = '/
+        ^(\s*)                               # Any leading whitespace.
+        <(?<tag_name>
+            h[1-6]                           # The opening tag...
+            (?=\s|>)                         # ...followed by a whitespace or >
+                                             # ?= means a "lookahead"
+        )
+        (?<before_class>                     # Any attributes prior to "class"
+            (?:                              # ?: is a "non-capturing group"
+                (?!\sclass=")[^>]            # Match all characters until ">" except when
+                                             # the next character sequence is \sclass="
+                                             # ?! is a "negative lookahead"
+            )*
+        )
+        (?:\s*class="(?<class_name>[^"]+)")? # The class attribute, if any
+        (?<after_class>[^>]*?)               # The rest of the tag
+        >                                    # The closing tag
+    /xm';
```

Comment:
I believe we should do this for all blocks and automagically add the `wp-block-{blockName}` class when rendering the pages on the server.

Reply:
Agreed! Let's ship this PR and explore it in a follow-up

Reply:
The follow-up: https://github.com/WordPress/gutenberg/pull/42269